subsys/mpsl/mpsl_init.c

```diff
@@ -124,6 +128,171 @@ static uint8_t m_config_clock_source_get(void)
 #endif
 }
 
+#if IS_ENABLED(CONFIG_MPSL_FEM)
+
+#define MPSL_FEM_NRF21540_GPIO_POLARITY_GET(dt_property) \
+	((DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), dt_property) & GPIO_ACTIVE_LOW) ? false : true)
+
+#define MPSL_FEM_GPIO_INVALID_PIN        0xFFU
+#define MPSL_FEM_GPIOTE_INVALID_CHANNEL  0xFFU
+#define MPSL_FEM_DISABLED_GPIO_CONFIG_INIT \
+	.enable       = false, \
+	.active_high  = true, \
+	.gpio_pin     = MPSL_FEM_GPIO_INVALID_PIN, \
+	.gpiote_ch_id = MPSL_FEM_GPIOTE_INVALID_CHANNEL
+
+#endif
+
+#if IS_ENABLED(CONFIG_MPSL_FEM_NRF21540_GPIO)
+static int fem_nrf21540_gpio_configure(void)
+{
+	/* FEM configuration requires gpiote and ppi channels.
+	 * Currently there is no reliable common method to dynamically
+	 * allocate such channels. FEM module needs only "some" channels
+	 * to use whichever they are, but FEM needs them for exclusive use
+	 * and does not enable them immediately.
+	 *
+	 * When common api to assign gpiote and ppi channels is available
+	 * current solution based on macros coming from Kconfig should be
+	 * reworked.
+	 */
+
+#if !DT_NODE_EXISTS(DT_NODELABEL(nrf_radio_fem))
+#error Node with label 'nrf_radio_fem' not found in the devicetree.
+#endif
+
+	mpsl_fem_nrf21540_gpio_interface_config_t cfg = {
+		.fem_config = {
+			.pa_time_gap_us  =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_settle_time_us),
+			.lna_time_gap_us =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_en_settle_time_us),
+			.pdn_settle_us   =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), pdn_settle_time_us),
+			.trx_hold_us     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), trx_hold_time_us),
+			.pa_gain_db      =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), tx_gain_db),
+			.lna_gain_db     =
+				DT_PROP(DT_NODELABEL(nrf_radio_fem), rx_gain_db)
+		},
+		.pa_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), tx_en_gpios)
+			.enable       = true,
+			.active_high  = MPSL_FEM_NRF21540_GPIO_POLARITY_GET(tx_en_gpios),
+			.gpio_pin     = DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), tx_en_gpios),
+			.gpiote_ch_id = CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_TX_EN
+#else
+			MPSL_FEM_DISABLED_GPIO_CONFIG_INIT
+#endif
+		},
+		.lna_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), rx_en_gpios)
+			.enable       = true,
+			.active_high  = MPSL_FEM_NRF21540_GPIO_POLARITY_GET(rx_en_gpios),
+			.gpio_pin     = DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), rx_en_gpios),
+			.gpiote_ch_id = CONFIG_MPSL_FEM_NRF21540_GPIO_GPIOTE_RX_EN
+#else
+			MPSL_FEM_DISABLED_GPIO_CONFIG_INIT
+#endif
+		},
+		.pdn_pin_config = {
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+			.enable       = true,
```

Comment:
Can the PDN management be made configurable? E.g. in the nRF21540 "mixed-mode" control, where PA/LNA is controlled from GPIOs and some configuration fields such as TX_GAIN are set through SPI, it will be the application who will control the PDN to avoid clearing the configured settings after each Radio transaction. This will be the case in the RF testing samples such DTM.

Reply:
Assuming the described feature request for PDN is valid only together with nRF21540-mixed-mode avaliability - 
I can see it this way that with nRF21540-mixed-mode a new separate FEM initialization method will be added in mpsl_init.c. Then a new variant of the PDN pin configuration will be taken into account, most likely using an additional flag from the Kconfig.

Reply:
Agree, this is a good approach.

Reply:
But I think such additional mode could be implemented later, out of scope of this PR. Right?

Reply:
Yes, I meant adding "mixed-mode" and a new way to control pin PDN in the next step in the future so it is beyond the scope of this PR.